clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -302,6 +303,150 @@ record_filter_t::write_trace_entries(per_shard_t *shard,
+                per_shard->input_entry_count - per_shard->input_count_at_ordinal == 1)
+                output = false;
+            break;
+        case TRACE_MARKER_TYPE_CPU_ID:
+            if (output)
+                per_shard->last_cpu_id = entry.addr;
+            // We insert our own start-of-chunk cpuid.
+            if (per_shard->archive_writer &&
+                per_shard->input_entry_count - per_shard->input_count_at_ordinal == 2)
+                output = false;
+            break;
+        case TRACE_MARKER_TYPE_PHYSICAL_ADDRESS:
+        case TRACE_MARKER_TYPE_PHYSICAL_ADDRESS_NOT_AVAILABLE:
+            if (!output && per_shard->archive_writer) {
+                // These markers need to be repeated across chunks, yet even raw2trace
+                // doesn't support this yet: so we bail on it here too.
+                return "Removing physical address markers from archive output is not yet "
+                       "supported";
+            }
+            break;
+        }
+    }
+    return "";
+}
+
+std::string
+record_filter_t::process_chunk_encodings(per_shard_t *per_shard, trace_entry_t &entry,
+                                         bool output)
+{
+    if (!per_shard->archive_writer ||
+        !is_any_instr_type(static_cast<trace_type_t>(entry.type)))
+        return "";
+    if (!per_shard->last_encoding.empty()) {
+        per_shard->pc2encoding[entry.addr] = per_shard->last_encoding;
+        // Disable the just-delayed encoding output below if this is
+        // what used to be a new-chunk encoding.
+        if (per_shard->cur_chunk_pcs.find(entry.addr) != per_shard->cur_chunk_pcs.end()) {
+            VPRINT(this, 3, "clearing new-chunk last encoding @pc=0x%zx\n", entry.addr);
+            per_shard->last_encoding.clear();
+        }
+    } else if (output) {
+        // Insert the cached encoding if this is the first instance of this PC
+        // (without an encoding) in this chunk, unless the user is removing all encodings.
+        // XXX: What if there is a filter removing all encodings but only
+        // to the stop point, so a partial remove that does not change
+        // the filetype?  For now we do not support that, and we re-add
+        // encodings at chunk boundaries regardless.
+        if (TESTANY(OFFLINE_FILE_TYPE_ENCODINGS, per_shard->filetype) &&
+            per_shard->cur_chunk_pcs.find(entry.addr) == per_shard->cur_chunk_pcs.end()) {
+            if (per_shard->pc2encoding.find(entry.addr) == per_shard->pc2encoding.end()) {
+                return "Missing encoding for PC " + std::to_string(entry.addr) +
+                    " at input entry " + std::to_string(per_shard->input_entry_count);
+            }
+            VPRINT(this, 3,
+                   "output new-chunk encoding chunk=%" PRIu64 " ref=%" PRIu64 "\n",
+                   per_shard->chunk_ordinal, per_shard->cur_refs);
+            if (!write_trace_entries(per_shard, per_shard->pc2encoding[entry.addr])) {
```

Comment:
Hmm, it is possible they could be merged, but at first glance I'm afraid of complications with other readers of last_encoding: I think it may not end up simpler.

Yes had to add the clear to avoid a double encoding which showed up in a new test (or may have been old test but after one of the review changes).